From 5da723612243a0af8cfcc3aa4856a6aaf4ba4dc7 Mon Sep 17 00:00:00 2001
From: Xianzhong <b07117@freescale.com>
Date: Tue, 13 Aug 2013 23:16:24 +0800
Subject: [PATCH 03/14] ENGR00274782 fixed gpu crash when baseAddress is not 0
 or 2G

The baseAddress of contiguousVidMem is the actual physical address
which is not subtracted by gpu baseAddress, but the allocated physical address
has been subtracted by gpu baseAddress in gckVIDMEM_Lock,
so the invalid offset is produced and used to calculate the logical address.

Signed-off-by: Xianzhong <b07117@freescale.com>
Acked-by: Shawn Guo
---
 drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.c
index 22c4071..de82c12 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_linux.c
@@ -332,9 +332,16 @@ gckKERNEL_MapVideoMemoryEx(
         else
 #endif
         {
+            gctUINT32 baseAddress = 0;
+
+            if (Kernel->hardware->mmuVersion == 0)
+            {
+                gcmkONERROR(gckOS_GetBaseAddress(Kernel->os, &baseAddress));
+            }
+
             gcmkVERIFY_OK(
                 gckHARDWARE_SplitMemory(Kernel->hardware,
-                                        device->contiguousVidMem->baseAddress,
+                                        device->contiguousVidMem->baseAddress - baseAddress,
                                         &pool,
                                         &base));
         }
-- 
1.9.1

